File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        m = res = l = r = 0
        d = {}
        le = len(s)
        while r < le:
            d[s[r]] = d.get(s[r], 0 ) + 1
            m = max(m, d[s[r]])

            while (r-l+1) - m > k:
                d[s[l]]-=1
                l+=1
            res = max(res, r - l + 1)
            r+=1
        return res


__import__('atexit').register(lambda : open('display_runtime.txt','w').write('0'))

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        // From neetcode
        int max = 0;
        int res = 0;
        HashMap<Character, Integer> map = new HashMap();

        int l = 0, r = 0;
        while(r < s.length()){
            Character ch = s.charAt(r);
            map.put(ch, map.getOrDefault(ch , 0) + 1);
            max = Math.max(max, map.get(ch));

            // to be true it must be (r-l+1)<= k;
            while( (r - l + 1 ) - max > k){
                Character c = s.charAt(l);
                map.put(c, map.get(c) - 1);
                l++;
            }
            res = Math.max(res, r - l + 1);
            r++;
        }

        return res;
    }
}